backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/EnumClassLowering.kt

```diff
@@ -158,9 +188,84 @@ internal class EnumUsageLowering(val context: Context)
+    }
+
+    override fun visitBlock(expression: IrBlock): IrExpression {
+        if (!shouldLower(expression)) {
+            return super.visitBlock(expression)
+        }
+        // Will be initialized only when we found a branch that compares
+        // subject with compile-time known enum entry
+        val ordinalVariable: IrVariable by lazy {
+            val variable = createOrdinalVariable(expression)
+            expression.statements.add(1, variable)
+            variable
+        }
+        val areEqualByValue = context.ir.symbols.areEqualByValue.first {
+            it.owner.valueParameters[0].type == context.builtIns.intType
+        }
+        val whenExpr = expression.statements[1] as IrWhen
+        whenExpr.branches.filter { it.condition is IrCall }.forEach {
+            val eqEqCall = it.condition as IrCall
+            val entry = eqEqCall.getArguments()[1].second as? IrGetEnumValue
+            if (entry != null) {
+                val entryOrdinal = EnumEntryEnumerator.getOrdinal(context, entry.descriptor)
+                // replace condition with trivial comparison of ordinals
+                it.condition = IrCallImpl(eqEqCall.startOffset, eqEqCall.endOffset, areEqualByValue).apply {
+                    putValueArgument(0, IrConstImpl.int(entry.startOffset, entry.endOffset, context.builtIns.intType, entryOrdinal))
+                    putValueArgument(1, IrGetValueImpl(ordinalVariable.startOffset, ordinalVariable.endOffset, ordinalVariable.symbol))
+                }
+            }
+        }
+        // Process nested when constructs.
+        expression.transformChildrenVoid(this)
+        return expression
+    }
+
+    // Checks that irBlock satisfies all constrains of this lowering.
+    // 1. Block's origin is WHEN
+    // 2. Subject of `when` is variable of enum type
+    private fun shouldLower(irBlock: IrBlock): Boolean {
+        if (irBlock.origin != IrStatementOrigin.WHEN) {
+            return false
+        }
+        // when-block should have two children: temporary variable and when itself.
+        assert(irBlock.statements.size == 2)
+        val tempVariable = irBlock.statements[0] as IrVariable
+        val enumClass = tempVariable.type.constructor.declarationDescriptor as? ClassDescriptor ?: return false
+        return enumClass.kind == ClassKind.ENUM_CLASS
+    }
+
+    private fun createOrdinalVariable(irBlock: IrBlock): IrVariable {
+        val tempVariable = irBlock.statements[0] as IrVariable
+        val ordinalPropertyGetter = context.ir.symbols.enum.getPropertyGetter("ordinal")!!
+        val getOrdinal = IrCallImpl(tempVariable.startOffset, tempVariable.endOffset, ordinalPropertyGetter).apply {
+            dispatchReceiver = IrGetValueImpl(tempVariable.startOffset, tempVariable.endOffset, tempVariable.symbol)
+        }
+        // Create temporary variable for subject's ordinal.
+        val ordinalDescriptor = IrTemporaryVariableDescriptorImpl(tempVariable.descriptor,
```

Comment:
Is `tempVariable.descriptor` really a containing declaration for this new variable?